test/Mono.Linker.Tests.Cases.Expectations/Assertions/RecognizedReflectionAccessPatternAttribute.cs

```diff
@@ -0,0 +1,20 @@
+﻿using System;
+
+namespace Mono.Linker.Tests.Cases.Expectations.Assertions
+{
+	[AttributeUsage (AttributeTargets.Class, AllowMultiple = true, Inherited = false)]
+	public class RecognizedReflectionAccessPatternAttribute : BaseExpectedLinkedBehaviorAttribute
+	{
+		public RecognizedReflectionAccessPatternAttribute (string sourceMethod, string reflectionMethod, string accessedItem)
```

Comment:
I think it would be nice to add the overloads that make maintaining and writing tests easier.  When these attributes use strings only it leads to tests that fall apart at the slightest refactoring.  And it deters us from writing tests with nested types and generics as those tend to complicate the string you need to craft.

For example, see https://github.com/mono/linker/blob/master/test/Mono.Linker.Tests.Cases.Expectations/Assertions/RemovedInterfaceOnTypeInAssemblyAttribute.cs

Supporting an overload where as much type information is supplied via `Type` rather than `string` does add additional work for supporting the new attribute, but it makes for a better and more maintainable test writing experience.

Using your assertion in the `MethodUsedViaReflection` test, it could become

```
	[RecognizedReflectionAccessPatternAttribute(typeof(MethodUsedViaReflection), nameof(MethodUsedViaReflection.Main), "()",
		typeof(Type), nameof(Type.GetMethod), "System.String,System.Reflection.BindingFlags",
		typeof(MethodUsedViaReflection), nameof(MethodUsedViaReflection.OnlyCalledViaReflection), "()")]
```

Or even
```
	[RecognizedReflectionAccessPatternAttribute(typeof(MethodUsedViaReflection), nameof(MethodUsedViaReflection.Main),new Type[0],
		typeof(Type), nameof(Type.GetMethod), new [] {typeof(BindingFlags)},
		typeof(MethodUsedViaReflection), nameof(MethodUsedViaReflection.OnlyCalledViaReflection), new Type[0])]
```

The easiest overload would be
```
		public RecognizedReflectionAccessPatternAttribute (Type sourceMethodType, string sourceMethodName, string sourceMethodSignature,
			Type reflectionMethodType, string reflectionMethodName, string reflectionMethodSignature,
			Type accessedItemType, string accessItemMemberName, string accessedItemSignature = null)
		{
		}
```
There should be enough logic in `ResultChecker` to make supporting this easy since other attributes do similar things.

Going 1 step further we could let you define the parameters by type as well.  No other attributes allow this today (I don't know why, maybe I just never thought of it).  I'd say going this far is optional but it would be cool to have.  I was thinking null could mean no signature (i.e. look for a Field or property to assert). 

```		
		public RecognizedReflectionAccessPatternAttribute (Type sourceMethodType, string sourceMethodName, Type[] sourceMethodSignature,
			Type reflectionMethodType, string reflectionMethodName, Type[] reflectionMethodSignature,
			Type accessedItemType, string accessItemMemberName, Type[] accessedItemSignature = null)
		{
		}


```

Reply:
Thanks for the suggestion. I took it one step further and moved the attributes to the method which contains the pattern. This way the first set of parameters is not needed anymore. Makes it much cleaner.

Makes detecting when to turn on the test recorder a bit trickier, but should work well. If we need to make it work across assemblies, then we would have to either tweak that logic or introduce an explicit "turn on" attribute. But I don't see that happening any time soon (if ever).